msal_extensions/osx.py

```diff
@@ -0,0 +1,315 @@
+# pylint: disable=duplicate-code
+
+"""Implements a macOS specific TokenCache, and provides auxiliary helper types."""
+
+import os
+import ctypes as _ctypes
+
+OS_RESULT = _ctypes.c_int32
+
+
+class KeychainError(OSError):
+    """The parent of all exceptions that may be thrown by the functions implemented inside of the
+    Keychain type."""
+    def __init__(self, exit_status):
+        super(KeychainError, self).__init__()
+        self.exit_status = exit_status
+        # TODO: pylint: disable=fixme
+        #  use SecCopyErrorMessageString to fetch the appropriate message here.
+        self.message = \
+            '{} ' \
+            'see https://opensource.apple.com/source/CarbonHeaders/CarbonHeaders-18.1/MacErrors.h'\
+                .format(self.exit_status)
+
+
+class KeychainAccessDeniedError(KeychainError):
+    """The exceptions that's raised when a Keychain is set to be queried, but the current process
+    has not been give permission to access it, and a query comes anyway."""
+    EXIT_STATUS = -128
+
+    def __init__(self):
+        super(KeychainAccessDeniedError, self).__init__(KeychainAccessDeniedError.EXIT_STATUS)
+
+
+class NoSuchKeychainError(KeychainError):
+    """The exception that's raised when a Keychain is set to be queried, but it is non-existent, and
+    a query comes anyway."""
+    EXIT_STATUS = -25294
+
+    def __init__(self, name):
```

Comment:
Following the conversation above, the tactical suggestion here would be change this constructor to:

```python
def __init__(exit_status, name):
    super(NoSuchKeychainError, self).__init__(exit_status)
    self.name = name
```

You may wonder, "wouldn't this constructor expose the internal const EXIT_STATUS and then makes it a burden for the caller to provide such value"? The fact is, the caller of this constructor is our own code `_construct_error()`, which would already need to know the mapping between those exit_status and their corresponding classes anyway.

Regarding to downstream app developers, they won't need to call these constructors, and more importantly, we do not (and can not) completely shield them from knowing the exit_status anyway. They still need to know them and then look it up from Apple's source.

Reply:
Fixed in 2962bd8d69a24544a0f6f725e7e0e89cb16fbcea

Reply:
Resolving this conversation after [offline discussion](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/15#issuecomment-505079762)